### PR TITLE
ENH: sparse.linalg: An efficient implementation of GMRES method

### PR DESCRIPTION
**Pull request by zhaog6**
#### Reference issue

Closes and Ref. gh-13874

#### What does this implement/fix?

The PR reimplemented GMRES algorihm, the code in the PR is more efficient than the original version. Compared with the previous PR (#13874), this PR has too many updates and adjustments, hope you like it.

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @zhaog6. I see it's still draft; once it's ready for review it'd be great to see a short summary of where to focus for review and some asv benchmarks.

**Comment by zhaog6** [comment]
> Thanks @zhaog6. I see it's still draft; once it's ready for review it'd be great to see a short summary of where to focus for review and some asv benchmarks.

Thank you @rgommers for your attention to the algorithm. On my desktop and laptop, the performance of the new algorithm is about ~ 2 times higher than that of the original algorithm. The latest asv benchmark performance in SciPy  is as follows: (`Ubuntu 20.04.3, scipy.__version__=1.7.3, numpy.__version__=1.21.4, sys.version_info(major=3, minor=8, micro=10, releaselevel='final', serial=0)`)

**======= The Original GMRES ======**

2021-12-03 16-30-38 的屏幕截图

**====== The New GMRES ======**

2021-12-03 16-00-54 的屏幕截图

Using `python runtests.py -s sparse` is also fine on my machines, but I can't reproduce the errors under Windows platform in CI/CR of GitHub. So I set it as a draft for the time being.

File: scipy/sparse/linalg/_isolve/gmresk.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    A preconditioner ``P`` is chosen such that ``P`` is close to `A` but easy
+    to solve for. The preconditioner parameter required by this routine is
+    ``M = P^-1``. The inverse should preferably not be calculated
+    explicitly.  Rather, use the following template to produce ``M``:
+
+    # Construct a linear operator that computes P^-1 @ x.
+    import scipy.sparse.linalg as spla
+    M_x = lambda x: spla.spsolve(P, x)
+    M = spla.LinearOperator((n, n), M_x)
+
```
[the remark this answers is not in the document]
Comment (zhaog6) [reply scipy/sparse/linalg/_isolve/gmresk.py:86]: Ok. Indeed, references were omitted, thank you

**Comment by zhaog6** [comment]
Dear Ralf @rgommers and Andrew @lobpcg ,
I encountered a troublesome issue, I accidentally deleted the forked SciPy library in my repository, all PR submitted before have been closed and cannot be opened again. On my local (laptop), these PR branch still exist. I re-forked SciPy library. But when I resubmit these PR branches to my forked SciPy, I found that I had to rebuild new PR instead of reopening the closed PR. I'd like to ask if there is any good way to restore or reopening these the closed PRs. Thanks a lot.

**Comment by rgommers** [comment]
Hi @zhaog6, that is unfortunate. I don't think there's a way to reopen these PRs. It's not a huge problem though, I suggest just opening new PRs, Cc'ing the reviewers of the old PRs, and porting over unaddressed review comments (if any).

**Comment by zhaog6** [comment]
Thank you so much. Ok, I try it

**Pull request closed without merging**